source/v5/mqtt3_to_mqtt5_adapter.c

```diff
@@ -1651,7 +1681,549 @@ static uint16_t s_aws_mqtt_client_connection_5_publish(
+            aws_array_list_get_at(&subscribe_op->subscriptions, &record, 0);
+
+            topic_filter = record->subscription_view.topic_filter;
+        }
+
+        if (suback->reason_code_count > 0) {
+            granted_qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[0]);
+        }
+
+        (*subscribe_op->on_suback)(
+            &adapter->base,
+            subscribe_op->base.id,
+            &topic_filter,
+            granted_qos,
+            error_code,
+            subscribe_op->on_suback_user_data);
+    }
+
+    if (subscribe_op->on_multi_suback != NULL) {
+        AWS_VARIABLE_LENGTH_ARRAY(
+            struct aws_mqtt_topic_subscription, multi_sub_subscription_buf, suback->reason_code_count);
+        AWS_VARIABLE_LENGTH_ARRAY(
+            struct aws_mqtt_topic_subscription *, multi_sub_subscription_ptr_buf, suback->reason_code_count);
+        struct aws_mqtt_topic_subscription *subscription_ptr =
+            (struct aws_mqtt_topic_subscription *)multi_sub_subscription_buf;
+
+        struct aws_array_list multi_sub_list;
+        aws_array_list_init_static(
+            &multi_sub_list,
+            multi_sub_subscription_ptr_buf,
+            suback->reason_code_count,
+            sizeof(struct aws_mqtt_topic_subscription *));
+
+        size_t subscription_count = aws_array_list_length(&subscribe_op->subscriptions);
+
+        for (size_t i = 0; i < suback->reason_code_count; ++i) {
+            struct aws_mqtt_topic_subscription *subscription = subscription_ptr + i;
+            AWS_ZERO_STRUCT(*subscription);
+
+            subscription->qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[i]);
+
+            if (i < subscription_count) {
+                struct aws_mqtt_subscription_set_subscription_record *record = NULL;
+                aws_array_list_get_at(&subscribe_op->subscriptions, &record, i);
+
+                subscription->topic = record->subscription_view.topic_filter;
+                subscription->on_publish = record->subscription_view.on_publish_received;
+                subscription->on_publish_ud = record->subscription_view.callback_user_data;
+                subscription->on_cleanup = record->subscription_view.on_cleanup;
+            }
+
+            aws_array_list_push_back(&multi_sub_list, &subscription);
+        }
+
+        (*subscribe_op->on_multi_suback)(
+            &adapter->base,
+            subscribe_op->base.id,
+            &multi_sub_list,
+            error_code,
+            subscribe_op->on_multi_suback_user_data);
+    }
+
+    aws_mqtt3_to_mqtt5_adapter_operation_table_remove_operation(
+        &subscribe_op->base.adapter->operational_state, subscribe_op->base.id);
+}
+
+static int s_validate_adapter_subscribe_options(const struct aws_mqtt3_to_mqtt5_adapter_subscribe_options *options) {
+    if (options->subscription_count == 0) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
```

Comment:
`s_aws_mqtt_client_connection_5_unsubscribe` runs the topics through `aws_mqtt_is_valid_topic_filter` but I don't think subscribe or subscribe multiple does. Maybe we can add it to those or here?

Reply:
That would result in a validation failure which would fail the overall request.  I'll add a test that does a invalid subscribe though.

Reply:
Actually, thinking about it more, that should be a synchronous failure on all of unsubscribe/subscribe/subscribe-multi because that's the way the 311 impl handles it.

Reply:
NM it already is a synchronous failure when the MQTT5 operation creation fails due to validation.